src/main/java/org/junitpioneer/jupiter/report/PioneerReportListener.java

```diff
@@ -0,0 +1,220 @@
+ */
+public class PioneerReportListener implements TestExecutionListener {
+
+	TestPlan storedTestPlan;
+	PioneerReport pioneerReport;
+
+	ConcurrentHashMap<String, List<Property>> propertyCache = new ConcurrentHashMap<>();
+	ConcurrentHashMap<String, String> testStatusCache = new ConcurrentHashMap<>();
+
+	@Override
+	public void reportingEntryPublished(TestIdentifier testIdentifier, ReportEntry entry) {
+		String id = testIdentifier.getUniqueId();
+
+		// Read value from published entry
+		for (Map.Entry<String, String> singleEntry : entry.getKeyValuePairs().entrySet()) {
+			Property property = new Property(singleEntry.getKey(), singleEntry.getValue());
+			// Add to cache
+			if (propertyCache.containsKey(id)) {
+				propertyCache.get(id).add(property);
+			} else {
+				List<Property> currentProperties = new ArrayList<>();
+				currentProperties.add(property);
+				propertyCache.put(id, currentProperties);
+			}
+		}
+	}
+
+	@Override
+	public void executionFinished(TestIdentifier testIdentifier, TestExecutionResult testExecutionResult) {
+		if (testIdentifier.isTest()) {
+			// Store result in cache
+			testStatusCache.put(testIdentifier.getUniqueId(), testExecutionResult.getStatus().toString());
+		}
+	}
+
+	@Override
+	public void testPlanExecutionFinished(TestPlan testPlan) {
+		// Store plan to access it from methods
+		this.storedTestPlan = testPlan;
+		pioneerReport = new PioneerReport();
+
+		Set<TestIdentifier> allRoots = testPlan.getRoots();
+
+		for (TestIdentifier identifier : allRoots) {
+			if (identifier.isContainer()) {
+				pioneerReport.getTestcontainer().add(processContainer(identifier));
+			} else if (identifier.isTest()) {
+				pioneerReport.getTestcase().add(processTest(identifier));
+			}
+		}
+		writeReport(pioneerReport);
+	}
+
+	/**
+	 * Converts an {@code TestIdentifier} object into a {@code Testcontainer} for the report.
+	 *
+	 * @param container Object to be converted into {@code Testcontainer}
+	 * @return Created object
+	 */
+	Testcontainer processContainer(TestIdentifier container) {
+		Testcontainer testcontainer = new Testcontainer(container.getDisplayName());
+
+		// Add properties, if there are any
+		String uniqueId = container.getUniqueId();
+		Properties allProps = getProperties(uniqueId);
+
+		if (!allProps.getProperty().isEmpty()) {
+			testcontainer.properties = allProps;
+		}
+
+		Set<TestIdentifier> allChildren = storedTestPlan.getChildren(uniqueId);
+
+		for (TestIdentifier identifier : allChildren) {
+			if (identifier.isContainer()) {
+				testcontainer.getTestcontainer().add(processContainer(identifier));
+			} else if (identifier.isTest()) {
+				testcontainer.getTestcase().add(processTest(identifier));
+			}
+		}
+
+		return testcontainer;
+	}
+
+	/**
+	 * Converts an {@code TestIdentifier} object into a {@code Testcase} for the report.
+	 *
+	 * @param test Object to be converted into {@code Testcase} element
+	 * @return Created object
+	 */
+	Testcase processTest(TestIdentifier test) {
```

Comment:
Shouldn't this method be `private`?

Reply:
I don't create private methods, so tests are able to access them.